src/NuGet.Core/NuGet.Protocol/Plugins/Connection.cs

```diff
@@ -0,0 +1,263 @@
+        public Connection(IMessageDispatcher dispatcher, ISender sender, IReceiver receiver, ConnectionOptions options)
+        {
+            if (dispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(dispatcher));
+            }
+
+            if (sender == null)
+            {
+                throw new ArgumentNullException(nameof(sender));
+            }
+
+            if (receiver == null)
+            {
+                throw new ArgumentNullException(nameof(receiver));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            MessageDispatcher = dispatcher;
+            _sender = sender;
+            _receiver = receiver;
+            Options = options;
+            _closeEvent = new TaskCompletionSource<object>();
+
+            MessageDispatcher.SetConnection(this);
+        }
+
+        /// <summary>
+        /// Disposes of this instance.
+        /// </summary>
+        public void Dispose()
+        {
+            if (_isDisposed)
+            {
+                return;
+            }
+
+            CloseAsync().GetAwaiter().GetResult();
+
+            GC.SuppressFinalize(this);
+
+            _isDisposed = true;
+        }
+
+        /// <summary>
+        /// Asynchronously connects and handshakes with a remote target.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A task that represents the asynchronous operation.</returns>
+        /// <exception cref="OperationCanceledException">Thrown if <paramref name="cancellationToken" /> is cancelled.</exception>
+        /// <exception cref="InvalidOperationException">Thrown if the method has been called already.</exception>
+        public async Task ConnectAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (State != ConnectionState.ReadyToConnect)
+            {
+                throw new InvalidOperationException(Strings.Plugin_ConnectionAlreadyStarted);
+            }
+
+            _state = (int)ConnectionState.Connecting;
+
+            _receiver.MessageReceived += OnMessageReceived;
+            _receiver.Faulted += OnFaulted;
+
+            using (var symmetricHandshake = new SymmetricHandshake(
+                this,
+                Options.HandshakeTimeout,
+                Options.ProtocolVersion,
+                Options.MinimumProtocolVersion))
+            {
+                await Task.WhenAll(_receiver.ConnectAsync(cancellationToken), _sender.ConnectAsync(cancellationToken));
+
+                _state = (int)ConnectionState.Handshaking;
+
+                ProtocolVersion = await symmetricHandshake.HandshakeAsync(cancellationToken);
+
+                if (ProtocolVersion == null)
+                {
+                    throw new ProtocolException(Strings.Plugin_HandshakeFailed);
```

Comment:
Change state here? Also there probably should be a generic error handler for other exceptions.